core/services/scan.go

```diff
@@ -0,0 +1,105 @@
+package services
+
+import (
+	"context"
+	"errors"
+
+	"github.com/kubescape/go-logger"
+	"github.com/kubescape/go-logger/helpers"
+	"github.com/kubescape/kubevuln/core/domain"
+	"github.com/kubescape/kubevuln/core/ports"
+	"go.opentelemetry.io/otel"
+)
+
+type ScanService struct {
+	sbomCreator    ports.SBOMCreator
+	sbomRepository ports.SBOMRepository
+	cveScanner     ports.CVEScanner
+	cveRepository  ports.CVERepository
+	armoPlatform   ports.Platform
+}
+
+var _ ports.ScanService = (*ScanService)(nil)
+
+func NewScanService(sbomCreator ports.SBOMCreator, sbomRepository ports.SBOMRepository, cveScanner ports.CVEScanner, cveRepository ports.CVERepository, armoPlatform ports.Platform) *ScanService {
+	return &ScanService{
+		sbomCreator:    sbomCreator,
+		sbomRepository: sbomRepository,
+		cveScanner:     cveScanner,
+		cveRepository:  cveRepository,
+		armoPlatform:   armoPlatform,
+	}
+}
+
+func (s *ScanService) GenerateSBOM(ctx context.Context, imageID string, workload domain.Workload) error {
+	ctx, span := otel.Tracer("").Start(ctx, "GenerateSBOM")
+	defer span.End()
+	if imageID == "" {
+		return errors.New("missing imageID")
+	}
+	sbom, err := s.sbomRepository.GetSBOM(ctx, imageID, s.sbomCreator.Version())
+	if err != nil {
+		return err
+	}
+	if sbom.Content != nil {
+		logger.L().Ctx(ctx).Warning("SBOM already generated", helpers.String("imageID", imageID))
+		return nil
+	}
+	sbom, err = s.sbomCreator.CreateSBOM(ctx, imageID, domain.RegistryOptions{})
+	if err != nil {
+		return err
+	}
+	// TODO add telemetry to Platform
+	return s.sbomRepository.StoreSBOM(ctx, sbom)
+}
+
+func (s *ScanService) Ready() bool {
+	return s.cveScanner.Ready()
+}
+
+func (s *ScanService) ScanCVE(ctx context.Context, instanceID string, imageID string, workload domain.Workload) error {
+	ctx, span := otel.Tracer("").Start(ctx, "ScanCVE")
+	defer span.End()
+	if instanceID == "" {
+		return errors.New("missing instanceID")
```

Comment:
question: how will users of a service be able to verify that an error they receive is about a missing instance ID and handle it appropriately?

If we create errors with `errors.New()`, clients of the function will have to compare the string contents of the error. This can get tedious very quick. If I’m correct, the current best practices are about defining custom errors and comparing against them with `errors.Is()`, as documented in the [Go blog](https://go.dev/blog/go1.13-errors) and [this](https://blog.logrocket.com/error-handling-golang-best-practices/) Logrocket blog.

Reply:
good point, let's do it later